07_future_promise/homework/notification_sender.cxx

```diff
@@ -17,4 +25,41 @@ int main()
 	std::string message = "build failed";
 
 	// ...
+
+	std::vector<std::thread> threads;
+
+	start = std::chrono::steady_clock::now();
+
+	// start threads
+	for ( std::string str : responsible_devs )
+	{
+		threads.push_back( std::thread([]( std::string strDev )
+			{
+				try
+				{
+					bool bResult = libmail::send_mail(strDev, "message");
+					std::stringstream ss;
+					ss << "> Sending mail to: " << strDev << " - ";
+					ss << ((bResult) ? "OK" : "FAIL");
+					ss << std::endl;
+					std::cout << ss.str();
+				}
+				catch( const std::exception& e )
+				{
+					std::stringstream ss;
+					ss << "> Sending mail to: <exception> " << e.what();
+					ss << std::endl;
+					std::cout << ss.str();
+				}	 				
+			}, str));
+	}
+
+	// join thread
+	std::for_each( std::begin( threads ), std::end( threads ), std::mem_fn( &std::thread::join ) );
```

Comment:
Wolałbym, aby wypisywanie było obsługiwane tylko przez główny wątek, a wątek ma tylko wysyłać maila, a wszelki status przekazać przez future / promise. Tutaj na join() wątek główny czeka, aż wszystkie wyślą maile i wypiszą co trzeba. Co więcej kolejność przyłączania wątków ma znaczenie. Jeśli 2, 3, 4 i 5 zostały zakolejkowane przed 1, a tutaj najpierw przyłączamy wątek 1, to całą reszta i tak poczeka na wątek 1 i mamy trochę zablokowane zasoby.

Reply:
Zrobione. Zobacz nowy pull request.